man/wtf.6.in
.\"
.\" wtf.7 -    man page for wtf and wtfindex.
.\" 
.\" (c) Thomas Sutton <thsutton>, 2003
.\"
.Dd August 24, 2003
.Os GAMES
.Dt WTF 6
.Sh NAME
.Nm wtf
.Nd translates acronyms and filename suffixes for you.
.br
.Nm wtfindex
.Nd builds string file indexes for
.Nm wtf .
.br
.Nm wtfdump
.Nd lists the contents of a 
.Nm wtf
database.
.Sh SYNOPSIS
.Nm 
.Op Fl a
.Op Ar is
.Ar pattern
.br
.Nm wtfindex
.Ar filename
.br
.Nm wtfdump
.Ar filename

.\" .Op Ar output.dat

.Sh DESCRIPTION
The
.Nm
program looks-up the definition of a term. It supports a number of definition
sources. In this version they are an acronyms database and a filename suffixes
database.

Like the 
.Nx 1.5
version of 
.Nm , this version will ignore an 
.Dq is 
given on the command line, allowing the more natural usage:
.Nm 
.Ar is
.Ar wtf .

Like
.Xr man ,
.Nm 
will display all matches found when the 
.Fl a
flag is given on the command line.

The 
.Nm wtfindex
program builds an index of the entries in
.Ar filename .
This index is written to
.Ar filename.dat .
.Ar filename
must contain a series of lines. Each line should consist of a key (the term
being defined) and a value (the definition) separated by a single tab
character, and should be terminated by a newline character. 

The 
.Nm wtfdump
program lists the entries in 
.Ar filename 
in ascending order.


.Sh FILES
.Bl -tag -width DATADIR -compact
.It Pa DATADIR
System-wide directory to search for data files.

.It Pa $HOME/.wtf
User specific directory to search for data files.

.Sh ENVIRONMENT
.Bl -tag -width WTFPATH -compact
.It Pa WTFPATH
Semi-colon delimited list of directories to be searched.

.Sh SEE ALSO
.Xr fortune 6 ,
.Xr whatis 1 ,
.Xr wtf 6 ,
.Xr sort 1

.Sh BUGS
The index files are not transportable across machines of different endian-ness
and/or with different sizes of 
.Ar unsigned long .

The command-line argument handling code is very rudimentary.

No doubt there are a great many more. If you find any, please squash them and
send the pieces to the author.

.Sh HISTORY
.Nm 
first appeared in 
.Nx 1.5 .
.br
This version of 
.Nm
was created on Slackware Linux 9.0.0

.Sh AUTHOR
.nf
Thomas Sutton
(thsutton@utas.edu.au)


NEWS
NEWS for wtf
============

20030824  - Version 0.0.4 [Public]
    + Bug fixes (Thanks to Igor Pechtchanski).
    + Added wtfdump to list databases.
	+ Uses $WTFPATH if it exists.
	+ Command line -a flag returns all matches found.

20030620 - Version 0.0.3 [Public]
    + Now searches user directory ($HOME/.wtf), if it exists, before the
      system-wide directory.

20030618 - Version 0.0.2 [Internal]
    + No longer breaks with unsorted text files.
    + Now searches all files (With an index file) in compile-time configured
      directory.
    + Still probably buggy as hell and not as fast as it could be :-)
    + A saner set of Makefiles.

20030616 - Version 0.0.1 [Public]
    + Basic functionality 

src/wtf.c
/*
 * wtf.c -    This program is a reimplementation of wtf(6), from NetBSD 1.5
 *            and later and Slackware 9.0 and later.
 * 
 * (c) Thomas Sutton <thsutton@utas.edu.au>, 2003.
 * This program is released under the terms of the GNU General Public License.
 *
 * CHANGES:
 *    2003-08-24    Thomas Sutton    <thsutton@utas.edu.au>
 *        wtf.c:    Implemented $WTFPATH and return all hits.
 */

#include <sys/types.h>
#include <dirent.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include "search.h"
#include "help.h"
#include "conf.h"

#define OPTION_ALL 1
#define OPTION_USER 2

int main(int argc, char **argv)
{
    char    *searchPath;                    /* For search path processing... */
    char    *cPath;
    int     tmp;
    DIR     *dirp = NULL;                    /* For directory reading... */
    struct     dirent *dirent_p = NULL;
    char     *target, *def, *dbfile_n,        /* For searching... */
            *file_n, *userdir_n;
    int file_n_l = 0;
    int options = 0;


    /* Handle the command line... 
     * Really, really badly. I need to look at getopt :-) */
    if ((argc < 2) || (argc > 4)) {
        fprintf(stderr, "%s\n", usage(argv[0]));
        exit(-1);
    }
    if (argc == 3) {
        if (strcmp(argv[1], "is") == 0) { 
            target = argv[2];
        } else if (strcmp(argv[1], "-a") == 0) {
            target = argv[2];
            options |= OPTION_ALL;
        } else {
            fprintf(stderr, "%s\n", usage(argv[0]));
            exit(-1);
        }
    } else if (argc == 4) {
        for (tmp=1;tmp<3;tmp++){
            if (strcmp(argv[tmp], "is") == 0) { 
            } else if (strcmp(argv[tmp], "-a") == 0) {
                options |= OPTION_ALL;
            } else {
                fprintf(stderr, "%s\n", usage(argv[0]));
                exit(-1);
            }
        }
        target = argv[3];
    } else {
        target = argv[1];
    }

            
    /* Get our search path. */
    searchPath = getenv("WTFPATH");
    tmp = 0;

    /* If we don't have one... */
    if (!searchPath) {
        /* Do we know the users home directory? */
        cPath = getenv("HOME");
        if (cPath)    tmp = strlen(cPath)+5+2; else tmp = 1;
        
        /* Home much memory is our search path?*/
        tmp += strlen(datadir_n);
        searchPath = (char*)malloc(tmp);
        
        /* Write the user data dir and system data dir into the new search
         * path...*/
        if (cPath) {
            snprintf(searchPath, tmp, "%s/.wtf:%s", cPath, datadir_n);
        } else {
            snprintf(searchPath, tmp, "%s", datadir_n);
        }
    }

    /* Walk through the search path performing the search...*/
    cPath = strtok(searchPath, ":");
    while (cPath) {
            /* Open the directory... */
            dirp = opendir(cPath);
            if (dirp) {
                /* skip . and .. */
                dirent_p = readdir(dirp);
                dirent_p = readdir(dirp);
                /* for each file... */
                while (((dirent_p = readdir(dirp)) != NULL)) {
                    file_n = dirent_p->d_name;
                    file_n_l = strlen(file_n);

                    /* If the file looks like an index...*/
                    if (strcmp(((dirent_p->d_name)+file_n_l-4), ".dat")==0) {
                        /* Assume the equiv text file exists... */
                        file_n[file_n_l-4] = 0;
                        file_n_l += strlen(cPath+2);
                        dbfile_n = malloc(file_n_l);
                        snprintf(dbfile_n, file_n_l, "%s/%s", cPath, file_n);

                        /* Search the database we have found for the target. */
                        def = search(dbfile_n, target);
                        free(dbfile_n);
                        if (def) {
                            /* Should this output target or the key from the
                             * db? */
                            printf("%s: %s\n",target,def);
                            free(def);
                            if (!(options & OPTION_ALL)) {
                                exit(0);
                            }
                        }
                    }
                }
            }
            
            /* get next token... */
            cPath = strtok(NULL,":");
    }

    /* Free searchPath if we created it. */
    if (tmp) free(searchPath);
}
